virt UEFI fix

cbosdo opened the pull request:
### What does this PR do?

Don't report an change if user provides `efi=True` and we already have a `loader` set in the XML definition of the VM.

The issue comes from the fact that libvirt transforms the `firmware='efi'` into a pair of `<loader>` and `<nvram>` elements with the automatically picked paths. When updating the VM if `efi=True` we need to assume the already defined loader and nvram are unchanged.

The PR also converts one of the `virt.update()` tests to pytest and split it into smaller more manageable test functions.

### What issues does this PR fix or reference?
Fixes: #59188

### Merge requirements satisfied?
- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

cbosdo committed:
virt: serial and console fixes

While reworking the tests, I figured out the serial and console update
one wasn't quite right and it has shown errors in the code too.

cbosdo committed:
Convert the virt test_update() test to pytest and split it

Convert the huge test_update() function to pytest. While at it, also
split it into smaller more manageable test functions.

cbosdo committed:
virt: fix update of efi=True

Libvirt uses efi=True as a flag to look for the proper UEFI loader and
nvram template. This means that on a subsenquent state apply efi=True
will trigger a definition update even if not needed.

In case efi=True is provided, don't udpate if loader and nvram values
are already set.

cbosdo committed:
virt: remove useless try/except

cbosdo commented:
rerun pr-ubuntu1804-py3-pytest

cbosdo commented:
@waynew I just rebased the PR on top of master after the network PR merge. You're good to review that one. Hopefully it shouldn't be that hard even given the loc changes: most of it is unit test moving/splitting.

s0undt3ch commented on tests/pytests/unit/modules/virt/test_helpers.py:
```diff
@@ -34,3 +34,11 @@ def assert_called(mock, condition):
     assert not condition and not mock.called or condition and mock.called
+
+
+def assertEqualUnit(actual, expected, unit="KiB"):
```
Could we covert this name away from camel case? We want to dissociate these tests from unittest `assert<Something>`

cbosdo replied:
sure, I'll change those

cbosdo replied:
Done. I also added a commit to change the same thing for the remaining calls in the `tests/unit/modules/test_virt.py` file to ease future tests migration.

cbosdo committed:
unCamelCase assertEqualUnit in virt module unit tests

In order to ease future virt module unit tests migration to pytest,
change the assertEqualUnit function to match the corresponding one in
pytest helpers.

waynew commented on salt/modules/virt.py:
```diff
@@ -4187,8 +4185,8 @@ def _skip_update(names):
         "interface": _skip_update(["interfaces", "nic_profile"]),
         "graphics": _skip_update(["graphics"]),
-        "serial": _skip_update(["serial"]),
-        "console": _skip_update(["console"]),
```
**question**: I'm guessing this is correct - just wanted to double check that `"serial"` and `"console"` should only be plural (have an S) in the list passed to `_skip_update`, correct? I see that it's not being updated in the `to_skip` dict - just the singular.

Is that expected?

cbosdo replied:
that's correct. The array contains `virt.update()` parameter names (which are plural), while the key is used in the `_diff_*_lists()` name and are singular.

waynew commented on salt/modules/virt.py:
```diff
@@ -2007,10 +2004,11 @@ def _handle_efi_param(boot, desc):
             parent_tag.remove(parent_tag.find("loader"))
             parent_tag.remove(parent_tag.find("nvram"))
-        return True
+            return True
```
**suggestion (non-blocking)**: I'd love if we could refactor this function a bit. The bool logic smells funny (`is False` and `is True` is a code smell), and we can do `parent_tag.find("loader")` up to 3 times in here.

I took a stab on the test clinic to try and come up with a reasonable refactor but I'm not sure I grok enough about it.

Anyway, I don't think it should block this PR at all, but it'd be awesome if it could get simplified at some point!

cbosdo replied:
I agree this one would need a refactoring. I'll try to come to it later, but not in this PR.

waynew reviewed (dismissed):
@cbosdo Thanks for the PR! It looks great - checked this out locally and ran the tests, +1 for migrating to pytests.

I left a question - if I understood correctly then this is :100: 

I also left a suggestion but I don't think there's any reason to hold this up.

cbosdo commented:
re-run all

cbosdo commented:
re-run archlts-py3-pytest

cbosdo commented:
re-run centos7-py3-tcp-pytest

cbosdo commented:
re-run centos8-py3-pytest

cbosdo commented:
re-run debian9-py3-pytest

cbosdo commented:
re-run debian10-py3-pytest

cbosdo commented:
re-run amazon2-py3-pytest

cbosdo commented:
re-run debian9-py3-pytest

cbosdo commented:
re-run centos7-py3-pytest

cbosdo commented:
re-run opensuse15-py3-pytest

cbosdo commented:
re-run ubuntu1604-py3-tcp-pytest

cbosdo commented:
re-run ubuntu1804-py3-pytest

cbosdo committed:
Fix virtual network generated DNS XML for SRV records

s0undt3ch commented on tests/pytests/unit/modules/virt/test_domain.py:
```diff
@@ -1060,3 +1064,944 @@ def test_update_nic_hostdev_nochange(make_mock_network, make_mock_vm, test):
+    assert ret["mem"]
+    assert ret["cpu"]
+    setxml = ET.fromstring(virt.libvirt.openAuth().defineXML.call_args[0][0])
+    assert "2" == setxml.find("vcpu").text
```
For improved diff readability of what pytest reports when there are failures, the asserts should be in the form of:
```python
assert what_we_got == what_we_expect
```

s0undt3ch replied:
The above applies to almost all of the newly added asserts

cbosdo replied:
oh, that was never clear to me which way it was... and I think we had both ways depending on the code. I'll fix those today

cbosdo replied:
@s0undt3ch I changed the ones introduced in this PR. To ease patches handing, I will change the other asserts in the virt pytests once the opened PRs are merged.

s0undt3ch replied:
Sounds good

s0undt3ch replied:
Thank You.

cbosdo committed:
Reverse all asserts in code introduced by PR#59189

Asserts need to have the actual value on the left hand side and the
expected one on the right hand side. Reverting these in the newly
converted code.

In order to ease backporting to openSUSE package, this commit will remain
separate.

cbosdo committed:
Merge branch 'master' into efi-fix

cbosdo committed:
Merge branch 'master' into efi-fix

cbosdo commented:
re-run all

cbosdo commented:
re-run ubuntu1604-py3-pycryptodome-pytest

cbosdo commented:
re-run macosxcatalina-py3-pytest

cbosdo committed:
Merge branch 'master' into efi-fix
